compose: Show one time banner to explain messages fading.

prakhar1144 opened the pull request:
When the compose box is open, we fade messages which the user is not replying to, to reduce the chance they send a message to a recipient they didn't intend to. Also, it reduces the visual/cognitive processing required to figure out where their message is going to go.

But, it's not necessarily clear to users that what the fading means, so this commit adds a one-time compose banner to explain what's going on the first time this comes up.

Fixes parts of #29076 


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->
---
**Screenshots and screen captures:**


<details><summary>Non Interleaved View</summary>
<img src="https://github.com/user-attachments/assets/b112e616-0849-4c7f-822f-97a549d24de3">
</img>
</details> 

<details><summary>Interleaved view </summary>
<img src="https://github.com/user-attachments/assets/38c9aec7-a92c-48ec-8b22-656ee937ef4e">
</img>
</details> 


---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
Question for @alya :

The issue says:
> In Zulip, you can send messages to a conversation without going there. You are composing a message to #{stream}>{topic}. Messages in other conversations are faded to avoid distraction. [Got it]

What does [Got it] represents here ? Button, right? Do we need it here, as it doesn't perform any action!

timabbott commented:
I think the idea for a "Got it" button is it'd hide the banner, just like the "x", but with a bit more explicit hinting that you're not going to get this banner again.

timabbott commented on zerver/lib/hotspots.py:
```diff
@@ -89,6 +89,9 @@ def to_dict(self) -> Dict[str, str]:
         name="intro_recent_view_modal",
     ),
+    OneTimeNotice(
+        name="messages_fading_banner",
```
Would `intro_message_fading` be a better name? I realize that we used the `_banner` / `_modal` suffice on previous ones, but I'm thinking that if we were to, in the future, change this from a banner to a modal, we probably wouldn't want to have to rename the field.

prakhar1144 replied:
Yeah makes sense from future perspective. Renamed it to `intro_messages_fading`

timabbott commented on web/src/compose_notifications.ts:
```diff
@@ -245,6 +248,50 @@ export function notify_messages_outside_current_search(messages: Message[]): voi
+            assert(topic_name !== undefined);
+            assert(stream_id !== undefined);
+
+            recipient = `#${stream_name} > ${topic_name}`;
```
I wonder if we should consider making a variant of `web/templates/inline_decorated_stream_name.hbs`  called something like `web/templates/inline_decorated_stream_with_topic.hbs`, that could at least allow styling these more nicely in the future (e.g. with the correct privacy icon)? 

@alya thoughts?

prakhar1144 replied:
I think that'd be nice. We'll have to address that independently as `#${stream_name} > ${topic_name}` is used in few more banners.

timabbott replied:
Sure, maybe let's open a "refactoring" label issue for it.

prakhar1144 replied:
I'll open an issue.

prakhar1144 commented:
Updates:

* Addressed review comments.

> I think the idea for a "Got it" button is it'd hide the banner, just like the "x", but with a bit more explicit hinting that you're not going to get this banner again.

* Added "Got it" button. See screenshots in PR description.

alya commented:
If I close the compose box without sending a message or pressing "Got it", the banner incorrectly appears again next time the compose box is open, regardless of context.

![CleanShot 2024-04-10 at 12 23 13](https://github.com/zulip/zulip/assets/2090066/16cf1a89-92c7-4a25-a9f6-0029f7b1499e)

alya commented:
Also, is it possible to update the conversation in the banner? If it's not easy to do, let's rewrite the banner text to not include it, to avoid situations like this:

![CleanShot 2024-04-10 at 12 22 39](https://github.com/zulip/zulip/assets/2090066/73a8f1f3-ddaf-45ea-97b9-5fafb04fe718)

alya commented:
I verified that the banner is correctly not triggered from Recent conversations.

alya commented:
The timing of the banner showing up when you click into the main compose area feels good to me.

prakhar1144 commented:
> If I close the compose box without sending a message or pressing "Got it", the banner incorrectly appears again next time the compose box is open, regardless of context.

Fixed it

> Also, is it possible to update the conversation in the banner? If it's not easy to do, let's rewrite the banner text to not include it, to avoid situations like this:

Made the changes -- posted a video in PR description

> I verified that the banner is correctly not triggered from Recent conversations.

As mentioned in issue: Show the first time the user would see a "go to conversation" button in a message view **(i.e., not in Inbox or Recent conversations)**, but only once the cursor is in the main compose area. -- Messages are not faded in such views.

alya commented:
@timabbott I think you can take a look next? I haven't re-tested.

timabbott commented on web/src/compose_notifications.ts:
```diff
@@ -245,6 +248,71 @@ export function notify_messages_outside_current_search(messages: Message[]): voi
+export function maybe_show_one_time_messages_fading_banner(): void {
+    const is_faded = $(".recipient_row").hasClass("message-fade");
+
+    if (is_faded) {
```
Use the early return pattern to avoid indenting so much code.

We can also easily adjust the selector to check if any of the faded messages are actually in the viewport?

prakhar1144 replied:
Done.

timabbott commented on web/src/compose_notifications.ts:
```diff
@@ -245,6 +248,71 @@ export function notify_messages_outside_current_search(messages: Message[]): voi
+            const user_ids = people.user_ids_string_to_ids_array(user_ids_string);
+            const names = people.get_display_full_names(user_ids).sort();
+
+            recipient = names.join(", ");
```
You can use `util.format_array_as_list` here to get a better result (it'll use "and" to make a nicer conjunction).

prakhar1144 replied:
![image](https://github.com/zulip/zulip/assets/56781761/65fea67e-2d6f-49bc-93b5-2f7ebfddc304)

timabbott commented on web/src/compose_notifications.ts:
```diff
@@ -245,6 +248,71 @@ export function notify_messages_outside_current_search(messages: Message[]): voi
+            const emails_string = compose_state.private_message_recipient();
+            const user_ids_string = people.emails_strings_to_user_ids_string(emails_string);
+            assert(user_ids_string !== undefined);
+            const user_ids = people.user_ids_string_to_ids_array(user_ids_string);
```
`compose_pm_pill.get_user_ids` might be able to get you this value much more directly.

timabbott replied:
Ahh, OK, makes sense!

timabbott commented on web/src/compose_actions.ts:
```diff
@@ -373,6 +373,7 @@ export function cancel(): void {
     $("#compose_close").hide();
     clear_box();
     compose_banner.clear_message_sent_banners();
+    compose_banner.clear_messages_fading_banner();
```
Do we also want to clear this banner when changing the compose recipient? Otherwise, the displayed recipient can become stale.

(I could also imagine a world where we reword it to avoid displaying the recipient, which can be quite long -- @alya thoughts?)

prakhar1144 replied:
We already addressed this: https://github.com/zulip/zulip/pull/29634#issuecomment-2048284693 ?

Note: The recipient are updated in the banner only when user clicks in compose message textarea. (Same rule as we do with other cases).

prakhar1144 commented:
Ready to review. Addressed the review comments.

timabbott commented on web/src/compose_notifications.ts:
```diff
@@ -245,6 +249,67 @@ export function notify_messages_outside_current_search(messages: Message[]): voi
+        return;
+    }
+
+    if (onboarding_steps.ONE_TIME_NOTICES_TO_DISPLAY.has("intro_messages_fading")) {
```
I think we should be doing an early return if that `has` is not true, to avoid slowing down the app building all the HTML above that we're not going to use.

timabbott replied:
Oh, but doing that would break the live-update support. Hmm.

prakhar1144 replied:
> if we decide we want to not mark it as "read" until you click "Got it"... which might be a better model.

https://chat.zulip.org/#narrow/stream/101-design/topic/explaining.20message.20fading/near/1781283

prakhar1144 replied:
We decided to go with the "Got it" approach.

timabbott commented on web/templates/compose_banner/messages_fading_banner.hbs:
```diff
@@ -0,0 +1,10 @@
+        You are composing a message to <z-link>{recipient}</z-link>.
+        Messages in other conversations are faded to avoid distraction.
+        {{#*inline "z-link"}}<a class="above_compose_banner_action_link" href="{{narrow_url}}">{{> @partial-block}}</a>{{/inline}}
+        {{/tr}}
```
Seeing this in action, it feels very wordy, and I don't like how the recipient is quite likely to be line-wrapped. I'll post a proposal in #design.

prakhar1144 replied:
Updated the PR with no recipients. Though we are still finalizing the exact text.

timabbott commented on web/src/compose_setup.js:
```diff
@@ -436,6 +447,7 @@ export function initialize() {
 
     $("textarea#compose-textarea").on("focus", () => {
         compose_recipient.update_placeholder_text();
+        compose_notifications.maybe_show_one_time_messages_fading_banner();
```
I don't know what the cause is, but the first time that I click on a message to open compose in an interleaved view, this doesn't appear.

prakhar1144 replied:
So, the reason was `respond_to_message()` flow, first opens the compose box then adds fading.

Fixed it by adding a call to this `maybe_....` function in `complete_starting_tasks()` with a comment.

prakhar1144 commented:
Updates in the recent push:

* Show different banner text when fading occurs in interleaved view vs non-interleaved view.
<details><summary>Non Interleaved</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/c15edb73-67e8-4ced-9636-cd9b95925a2d">
</img>
</details> 

<details><summary>Interleaved view</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/744a5423-d3ea-4e1f-b5f2-02ff2693d46f">
</img>
</details> 

* Mark the banner as read only when "Got it" button is clicked.

prakhar1144 commented:
Update:
* Now we have [two independent banners as discussed.](https://chat.zulip.org/#narrow/stream/101-design/topic/explaining.20message.20fading/near/1785586)

alya commented:
The way the banner text updates doesn't quite feel right. It seems to require clicking out of the main compose area and back into it again. So, e.g., if I put the cursor in the compose area, and change my view from interleaved to non-interleaved, the banner text remains the same. It changes if I then click into the topic box and then back into the main compose area.

I don't think it's the end of the world if we don't quite get this right, but maybe it's not too hard to fix?

prakhar1144 commented:
Addressed the above two review comments: https://github.com/zulip/zulip/pull/29634#issuecomment-2073274029 https://github.com/zulip/zulip/pull/29634#issuecomment-2073284046

Posted updated screenshot in PR description: https://github.com/zulip/zulip/pull/29634#issue-2227914199

Also, posted a design question in CZO

prakhar1144 commented:
Found another edge case, fixing...!

prakhar1144 commented:
Updated the PR as per [recent discussion in CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/explaining.20message.20fading/near/1890260) to keep one variant of banner in interleaved view for 9.0 release.

Screenshot updated in [PR description](https://github.com/zulip/zulip/pull/29634#issue-2227914199).

alya commented:
Works for me in manual testing!

timabbott commented:
Looks good to me but I'll defer merging until I've integrated some other migrations for numbering-related reasons on chat.zulip.org.

timabbott commented:
Merged, after renumbering the migration. Thanks @prakhar1144 for all the work on this!
